server/src/server.ts

```diff
@@ -79,14 +79,16 @@ let codeActionsFromDiagnostics: codeActions.filesCodeActions = {};
 // will be properly defined later depending on the mode (stdio/node-rpc)
 let send: (msg: p.Message) => void = (_) => {};
 
-let findBinaryDirPathFromProjectRoot = (
-  directory: p.DocumentUri // This must be a directory and not a file!
+let findBinaryFromProjectDirRec = (
```

Comment:
I find the logic inside this function difficult to follow, so error prone to any changes.
Looks like the caller has all the context, and this function does not stand on its own legs.
(This is only about style, not functionality).